tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java

```diff
@@ -288,25 +291,83 @@ private JsonNode getCombinedStatus() {
+            if (id.isEmpty())
+                throw new InvalidRequestException("Invalid empty ID in createTask request.");
+
+            TaskSpec spec = idAndSpec.getValue();
+            ManagedTask duplicateTask = this.tasks.get(id);
+            if (duplicateTask != null) {
+                if (!duplicateTask.spec.equals(spec))
+                    throw new RequestConflictException("Task ID " + id + " already " +
+                        "exists, and has a different spec " + duplicateTask.spec);
+                log.info("Task {} already exists with spec {}", id, spec);
+                if (tasks.size() == 1)
+                    return;
+                else
+                    it.remove();
+            }
+        }
+        for (Map.Entry<String, TaskSpec> idAndSpec: tasks.entrySet()) {
+            try {
+                ManagedTask createdTask = createTask(idAndSpec.getKey(), idAndSpec.getValue());
+                if (createdTask != null)
+                    managedTasks.add(createdTask);
+            } catch (Throwable e) {
+                log.info("Failed to create createTask(id={}, spec={}) error:", idAndSpec.getKey(), idAndSpec.getValue(), e);
+            }
+        }
+
+        boolean infoLogEnabled = managedTasks.size() < 5;
+        for (ManagedTask task: managedTasks) {
+            long delayMs = scheduleTask(task);
+            if (infoLogEnabled)
+                log.info("Created a new task {} with spec {}, scheduled to start {} ms from now.", task.id, task.spec, delayMs);
+            else
+                log.debug("Created a new task {} with spec {}, scheduled to start {} ms from now.", task.id, task.spec, delayMs);
+        }
+
+        log.info("Created {} tasks.", managedTasks.size());
+    }
+
+    private long scheduleTask(ManagedTask task) {
+        tasks.put(task.id, task);
+        long delayMs = task.startDelayMs(time.milliseconds());
+        task.startFuture = scheduler.schedule(executor, new RunTask(task), delayMs);
+        return delayMs;
+    }
+
+    /**
+     * @return a #{@link ManagedTask} instance or #{@code null} if there was an error creating the task
+     *  or if the task already exists
+     */
+    private ManagedTask createTask(String id, TaskSpec spec) throws Throwable {
         try {
-            executor.submit(new CreateTask(id, spec)).get();
+            return executor.submit(new CreateTask(id, spec)).get();
         } catch (ExecutionException e) {
             log.info("createTask(id={}, spec={}) error", id, spec, e);
             throw e.getCause();
+        } catch (Exception e) {
+            log.info("Failed to create a new task {} with spec {}: {}",
+                id, spec, e.getMessage());
+            return null;
         }
     }
 
     /**
      * Handles a request to create a new task.  Processed by the state change thread.
      */
-    class CreateTask implements Callable<Void> {
+    class CreateTask implements Callable<ManagedTask> {
```

Comment:
I don't think this `Callable` should return `ManagedTask`.  `ManagedTask` is an internal data structure that should only be accessed from the state change thread.